From a7592ca1818090e598b2ee1262c977ab9c696842 Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Wed, 22 Jan 2025 17:47:03 +0100
Subject: [PATCH] ASoC: stm32: sai: skip useless iterations on kernel rate loop

the frequency of the kernel clock must be greater than or equal to the
bitclock rate. When searching for a convenient kernel clock rate in
stm32_sai_set_parent_rate() function, it is useless to continue the loop
below bitclock rate, as it will result in a invalid kernel clock rate.
Change the loop output condition.

Fixes: b1d2e4067dc6 ("ASoC: stm32: sai: add stm32mp25 support")
Change-Id: I1cd7dd98ae8aef1fbdd0d7027fa41323e4cc0b75
Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/433832
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 sound/soc/stm/stm32_sai_sub.c | 34 +++++++++++++++++++++++++---------
 1 file changed, 25 insertions(+), 9 deletions(-)

--- a/sound/soc/stm/stm32_sai_sub.c
+++ b/sound/soc/stm/stm32_sai_sub.c
@@ -132,7 +132,8 @@ struct stm32_sai_sub_data {
 	struct snd_aes_iec958 iec958;
 	struct mutex ctrl_lock; /* protect resources accessed by controls */
 	spinlock_t irq_lock; /* used to prevent race condition with IRQ */
-	int (*set_sai_ck_rate)(struct stm32_sai_sub_data *sai, unsigned int rate);
+	int (*set_sai_ck_rate)(struct stm32_sai_sub_data *sai, unsigned int rate,
+			       unsigned int clk_ref);
 	void (*put_sai_ck_rate)(struct stm32_sai_sub_data *sai);
 };
 
@@ -144,6 +145,11 @@ enum stm32_sai_fifo_th {
 	STM_SAI_FIFO_TH_FULL,
 };
 
+enum stm32_sai_clk_ref {
+	STM_SAI_CLK_FS,
+	STM_SAI_CLK_MCLK,
+};
+
 static bool stm32_sai_sub_readable_reg(struct device *dev, unsigned int reg)
 {
 	switch (reg) {
@@ -379,7 +385,7 @@ static bool stm32_sai_rate_accurate(unsi
 }
 
 static int stm32_sai_set_parent_clk(struct stm32_sai_sub_data *sai,
-				    unsigned int rate)
+				    unsigned int rate, unsigned int clk_ref)
 {
 	struct platform_device *pdev = sai->pdev;
 	struct clk *parent_clk = sai->pdata->clk_x8k;
@@ -406,10 +412,10 @@ static void stm32_sai_put_parent_rate(st
 }
 
 static int stm32_sai_set_parent_rate(struct stm32_sai_sub_data *sai,
-				     unsigned int rate)
+				     unsigned int rate, unsigned int clk_ref)
 {
 	struct platform_device *pdev = sai->pdev;
-	unsigned int sai_ck_rate, sai_ck_max_rate, sai_curr_rate, sai_new_rate;
+	unsigned int sai_ck_rate, sai_ck_max_rate, sai_ck_min_rate, sai_curr_rate, sai_new_rate;
 	int div, ret;
 
 	/*
@@ -428,8 +434,18 @@ static int stm32_sai_set_parent_rate(str
 	else
 		sai_ck_max_rate = SAI_MAX_SAMPLE_RATE_8K * 256;
 
-	if (!sai->sai_mclk && !STM_SAI_PROTOCOL_IS_SPDIF(sai))
-		sai_ck_max_rate /= DIV_ROUND_CLOSEST(256, roundup_pow_of_two(sai->fs_length));
+	/* If the reference clock is the master clock, rate = mclk rate. otherwise rate = fs */
+	if (clk_ref == STM_SAI_CLK_MCLK) {
+		sai_ck_min_rate = rate;
+	} else {
+		if (STM_SAI_PROTOCOL_IS_SPDIF(sai)) {
+			sai_ck_min_rate = rate * 256;
+		} else {
+			sai_ck_min_rate = rate * sai->fs_length;
+			sai_ck_max_rate /= DIV_ROUND_CLOSEST(256,
+							     roundup_pow_of_two(sai->fs_length));
+		}
+	}
 
 	/*
 	 * Request exclusivity, as the clock is shared by SAI sub-blocks and by
@@ -472,7 +488,7 @@ static int stm32_sai_set_parent_rate(str
 		/* Try a lower frequency */
 		div++;
 		sai_ck_rate = sai_ck_max_rate / div;
-	} while (sai_ck_rate > rate);
+	} while (sai_ck_rate >= sai_ck_min_rate);
 
 	/* No accurate rate found */
 	dev_err(&pdev->dev, "Failed to find an accurate rate");
@@ -694,7 +710,7 @@ static int stm32_sai_set_sysclk(struct s
 		}
 
 		/* If master clock is used, configure SAI kernel clock now */
-		ret = sai->set_sai_ck_rate(sai, freq);
+		ret = sai->set_sai_ck_rate(sai, freq, STM_SAI_CLK_MCLK);
 		if (ret)
 			return ret;
 
@@ -1121,7 +1137,7 @@ static int stm32_sai_configure_clock(str
 	int ret;
 
 	if (!sai->sai_mclk) {
-		ret = sai->set_sai_ck_rate(sai, rate);
+		ret = sai->set_sai_ck_rate(sai, rate, STM_SAI_CLK_FS);
 		if (ret)
 			return ret;
 	}
